src/components/editor/block-selections/caret.js

```diff
@@ -0,0 +1,96 @@
+/**
+ * WordPress dependencies
+ */
+import classnames from 'classnames';
+import { applyFormat, registerFormatType } from '@wordpress/rich-text';
+
+const FORMAT_NAME = 'asblocks/caret';
+
+/**
+ * Applies given carets to the given record.
+ *
+ * @param {Object} record The record to apply carets to.
+ * @param {Array} carets The carets to apply.
+ * @return {Object} A record with the carets applied.
+ */
+export function applyCarets( record, carets = [] ) {
+	carets.forEach( ( caret ) => {
+		let { start, end, id } = caret;
+		const isCollapsed = start === end;
+		const isShifted = isCollapsed && end >= record.text.length;
+
+		if ( isShifted ) {
+			start = record.text.length - 1;
+		}
+
+		if ( isCollapsed ) {
+			end = start + 1;
```

Comment:
What if the caret is at the end? Wonder what `object: true` on the format object would do.

Reply:
In case of an object, it wouldn't need to be applied as a format, but instead be inserted at the caret position.

Reply:
Interesting, I didn't know how to use objects. Right now if it's at the end, I just shift it backward by 1 and display the border on the right instead of on the left.